File: src/main/java/com/my/admin/app/commerce/service/ProductService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.ProductDao;
import com.black.food.manager.model.api.response.ProductResponse;
import com.black.food.manager.model.api.response.ProductTypeResponse;
import com.black.food.manager.model.entity.Product;
import com.black.food.manager.model.transformer.ProductTransformer;
import com.black.food.manager.model.transformer.ProductTypeTransformer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class ProductService {

    private ProductDao productDao;

    private ExceptionService exceptionService;

    public Product findById(final Long id) {
        return productDao.findById(id).orElseThrow();
    }

    public Set<ProductResponse> productFilter(final Long id, final Long productTypeId, final String nameCode) {
        var products = productDao.productFilter(id, productTypeId, nameCode);

        products.stream()
                .findFirst().orElseThrow(() -> exceptionService.throwNotFoundException("No se encontraron productos para los filtros seleccionados."));

        return products
                .stream()
                .map(ProductTransformer.FROM_PRODUCT_TO_PRODUCT_RESPONSE)
                .sorted((p1, p2) -> p1.getName().toUpperCase().compareTo(p2.getName()))
                .collect(Collectors.toCollection(LinkedHashSet::new));
    }

    public Set<ProductResponse> findProductToppingsByMainDishId(final Long mainDishId) {
        return productDao.findProductToppingsByMainDishId(mainDishId);
    }

    public Set<ProductTypeResponse> findProductTypes() {
        var productTypeStream = productDao.findAllProductsTypes();

        productTypeStream.stream().findFirst().orElseThrow(() -> exceptionService.throwNotFoundException("No se encontraron tipos de productos"));

        return productTypeStream
                .stream()
                .map(ProductTypeTransformer.FROM_PRODUCT_TYPE_TO_PRODUCT_TYPE_RESPONSE)
                .collect(Collectors.toSet());
    }

    @Autowired
    public void setProductDao(final ProductDao productDao) {
        this.productDao = productDao;
    }

    @Autowired
    public void setExceptionService(final ExceptionService exceptionService) {
        this.exceptionService = exceptionService;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/ProductService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.ProductDAO
import com.bk.reactive.app.commerce.admin.my.exception.api.ApplicationException
import com.bk.reactive.app.commerce.admin.my.model.api.response.ProductResponse
import com.bk.reactive.app.commerce.admin.my.model.api.response.ProductTypeResponse
import com.bk.reactive.app.commerce.admin.my.model.mapper.ProductMapper.Companion.fromProductToProductResponse
import com.bk.reactive.app.commerce.admin.my.model.mapper.ProductTypeMapper.Companion.fromProductTypeToProductTypeResponse
import mu.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class ProductService(
    private val productDAO: ProductDAO,
    private val exceptionService: ExceptionService
) {
    private val logger = KotlinLogging.logger {}

    fun findById(id: String): Mono<ProductResponse> {
        return productDAO.findById(id)
            .map { fromProductToProductResponse(it) }
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error {
                    ApplicationException(message = it.localizedMessage, cause = Throwable(it.localizedMessage))
                }
            }
    }

    fun productFilter(id: String?, productTypeId: String?, nameCode: String?): Flux<ProductResponse> {
        return productDAO.findAll()
            .filter { product -> id == null || product.id == id }
            .filter { product -> productTypeId == null || product.productType.id == productTypeId }
            .switchIfEmpty(
                Mono.error {
                    ApplicationException(
                        message = "No se encontraron productos para los filtros seleccionados.",
                        code = HttpStatus.NOT_FOUND,
                        cause = Throwable()
                    )
                }
            ).map { fromProductToProductResponse(it) }
    }

    fun findProductToppingsByMainDishId(mainDishId: String): Flux<ProductResponse> {
        return productDAO.findProductToppingsByMainDishId(mainDishId)
            .switchIfEmpty(
                Mono.error { exceptionService.throwNotFoundException("No se encontraron toppings para el producto sellecionado") }
            ).map { fromProductToProductResponse(it) }
    }

    fun findProductTypes(): Flux<ProductTypeResponse> {
        return productDAO.findAllProductsTypes()
            .switchIfEmpty {
                Flux.error<ProductTypeResponse> {
                    exceptionService.throwNotFoundException("No se encontraron tipos de productos.")
                }
            }.map {
                fromProductTypeToProductTypeResponse(it)
            }
    }
}